util/network-devp2p/src/host.rs

```diff
@@ -1218,51 +1298,85 @@ impl IoHandler<NetworkIoMessage> for Host {
 	}
 }
 
-fn save_key(path: &Path, key: &Secret) {
+trait DiskEntity: FromStr {
+	fn path() -> &'static str;
+	fn desc() -> &'static str;
+	fn to_repr(&self) -> String;
+}
+
+impl DiskEntity for Secret {
+	fn path() -> &'static str {
+		"key"
+	}
+	fn desc() -> &'static str {
+		"key file"
+	}
+	fn to_repr(&self) -> String {
+		self.to_hex()
+	}
+}
+
+impl DiskEntity for Enr {
+	fn path() -> &'static str {
+		"enr"
+	}
+	fn desc() -> &'static str {
+		"Ethereum Node Record"
+	}
+	fn to_repr(&self) -> String {
+		self.to_base64()
+	}
+}
+
+fn save<E: DiskEntity>(path: &Path, entity: &E) {
 	let mut path_buf = PathBuf::from(path);
 	if let Err(e) = fs::create_dir_all(path_buf.as_path()) {
-		warn!("Error creating key directory: {:?}", e);
+		warn!("Error creating {} directory: {:?}", E::desc(), e);
 		return;
 	};
-	path_buf.push("key");
+	path_buf.push(E::path());
 	let path = path_buf.as_path();
 	let mut file = match fs::File::create(&path) {
 		Ok(file) => file,
 		Err(e) => {
-			warn!("Error creating key file: {:?}", e);
+			warn!("Error creating {}: {:?}", E::desc(), e);
 			return;
 		}
 	};
 	if let Err(e) = restrict_permissions_owner(path, true, false) {
 		warn!(target: "network", "Failed to modify permissions of the file ({})", e);
 	}
-	if let Err(e) = file.write(&key.to_hex().into_bytes()) {
-		warn!("Error writing key file: {:?}", e);
+	if let Err(e) = file.write(&entity.to_repr().into_bytes()) {
+		warn!("Error writing {}: {:?}", E::desc(), e);
 	}
 }
 
-fn load_key(path: &Path) -> Option<Secret> {
+fn load<E>(path: &Path) -> Option<E>
+where
+	E: DiskEntity,
+	<E as std::str::FromStr>::Err: std::fmt::Debug,
```

Comment:
```suggestion
	<E as std::str::FromStr>::Err: Debug,
```
I think this is just part of the prelude, no?